### PR TITLE
Add a new test for user/group resolution in nss modules

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

File: src/test/test-nss-users.c [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+#include "nss-util.h"
+#include "path-util.h"
+#include "parse-util.h"
+#include "stdio-util.h"
+#include "string-util.h"
+#include "strv.h"
+#include "tests.h"
+#include "user-util.h"
+
+size_t arg_bufsize = 1024;
```
Comment (poettering) [inline src/test/test-nss-users.c]: static?

File: src/test/test-nss-users.c [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+
+        if (argc > 2)
+                names = strv_copy(strv_skip(argv, 2));
+        else
+                names = strv_new("root",
+                                 NOBODY_USER_NAME,
+                                 "foo_no_such_user",
+                                 "0",
+                                 "65534");
+        if (!names)
```
Comment (poettering) [inline src/test/test-nss-users.c]: assert_se() here... it's a test, ENOMEM can#t really happen

File: src/test/test-nss-users.c [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+
+        test_setup_logging(LOG_INFO);
+
+        r = parse_argv(argc, argv, &modules, &names);
+        if (r < 0) {
+                log_error_errno(r, "Failed to parse arguments: %m");
+                return EXIT_FAILURE;
+        }
+
+        dir = dirname_malloc(argv[0]);
```
Comment (poettering) [inline src/test/test-nss-users.c]: maybe use path_extract_directory(), it's a lot safer

File: src/test/test-nss-users.c [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+        if (!dir)
+                return log_oom();
```
Comment (poettering) [inline src/test/test-nss-users.c]: in tests i'd catch oom via assert_se() really

File: src/test/test-nss-users.c [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+                assert_cc(sizeof(uid_t) == sizeof(uint32_t));
+                /* We use safe_atou32 because we don't want to refuse invalid uids. */
+                if (safe_atou32(*name, &uid) < 0)
+                        continue;
+
+                test_byuid(handle, module, uid);
+        }
+
+        log_info(" ");
+        dlclose(handle);
```
Comment (poettering) [inline src/test/test-nss-users.c]:
```c
_cleanup_(dlclosep) void *handle = NULL;
```

is a bit prettier

File: src/test/test-nss-users.c [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+        STRV_FOREACH(name, names)
+                test_byname(handle, module, *name);
+
+        STRV_FOREACH(name, names) {
+                uid_t uid;
+
+                assert_cc(sizeof(uid_t) == sizeof(uint32_t));
+                /* We use safe_atou32 because we don't want to refuse invalid uids. */
+                if (safe_atou32(*name, &uid) < 0)
+                        continue;
```
Comment (poettering) [inline src/test/test-nss-users.c:183]: parse failure are something i'd really log about

File: docs/ENVIRONMENT.md [hunk 7]
```diff
@@ -302,6 +297,14 @@ installed systemd tests:
 * `$SYSTEMD_SYSVRCND_PATH` — Controls where `systemd-sysv-generator` looks for
   SysV init script runlevel link farms.
 
+systemd tests:
+
+* `$SYSTEMD_TEST_DATA` — override the location of test data. This is useful if
+  a test executable is moved to an arbitrary location.
+
+* `$SYSTEMD_TEST_NSS_BUFSIZE` — size of scratch buffers for "reentrant"
+  functions.
```
Comment (poettering) [inline docs/ENVIRONMENT.md]: maybe mention this is about NSS? the env var suggests so at least

File: src/test/test-nss-hosts.c [hunk 8]
```diff
@@ -17,12 +17,15 @@
...
 #include "nss-test-util.h"
 #include "nss-util.h"
+#include "parse-util.h"
 #include "path-util.h"
 #include "stdio-util.h"
 #include "string-util.h"
 #include "strv.h"
 #include "tests.h"
 
+size_t arg_bufsize = 1024;
```
Comment (poettering) [inline src/test/test-nss-hosts.c]: static?

**Review by poettering — COMMENTED** [review]
looks good, just some minor comments

File: src/test/test-nss-users.c (hunk 6 above)
Comment (keszybz) [reply src/test/test-nss-users.c:183]: The idea is that you can specify some name arguments and some number arguments, and if it's a number it gets handled as uid, and if its a name, it gets handled as a name. So it shouldn't log.

**Comment by keszybz** [comment]
Updated with the suggested changes, except when commented above. Taking the liberty to set the green label.

**Pull request merged**